### ISSUE TITLE
Retrospective for April 2020 releases

### ISSUE THREAD

**Issue by sxa** [AdoptOpenJDK/TSC#151]
1. Testing was removed for release builds as well as nightlies _(Let's doc correct process)_
2. JDK8u mirroring failed (missing langtools directory)
3. JDK11/Windows/HotSpot build [failed on one build machine](https://github.com/AdoptOpenJDK/openjdk-build/issues/1667)
4. AIX JDK8 builds are [suffering glitches](https://github.com/AdoptOpenJDK/openjdk-infrastructure/issues/1279) (at least for nightly runs)
5. New skara mirror (JDK14+) missing a `git push --tags` in the [new mirror script](https://github.com/AdoptOpenJDK/openjdk-build/blob/master/git-hg/skaraMirror.sh#L124)
6. Windows JREs contained debug symbols and were [double the expected size](https://github.com/AdoptOpenJDK/openjdk-build/issues/1670) (11.0.7/hotspot already out, will need respin, can catch it for other releases)
7. Windows GA release was [not properly signed](https://github.com/AdoptOpenJDK/openjdk-build/issues/1694) after [new Azure build machines added](https://github.com/AdoptOpenJDK/openjdk-infrastructure/issues/1239)
8. Miscellaneous concerns about triage simplicity
9. Unclear use of `adoptBuildNumber` in the piplines. Apparently defaults to `1`. [Raised](https://github.com/AdoptOpenJDK/openjdk-build/issues/1696) Also caused [issue with API](https://github.com/AdoptOpenJDK/openjdk-api-v3/issues/209#issuecomment-631774151)

**Comment by andrew-m-leonard** [comment]
2. update on jdk8u mirror issue, it "seemed" as RedHat merged jdk8u252-b09 early evening, the nightly mirror job merged the tag jdk8u252-b09 already, this then seemed to impact the move-to-next-tag action as the tag should not have existed. I suspect before we ran the move to next tag as the first thing after RedHat merged it

**Comment by smlambert** [comment]
Based on history, we should properly address the longstanding failures (as per https://github.com/AdoptOpenJDK/TSC/issues/152#issuecomment-614107570), preferably fix them, but at very least exclude.

**Comment by sxa** [comment]
Also related to that we have various issues from efore for making things easier for releasing:
- [Build script updates](https://github.com/AdoptOpenJDK/openjdk-build/issues/1613)
- [Updates to cover docker releases](https://github.com/AdoptOpenJDK/openjdk-build/issues/1613)
- [Updates for new major releases (not relevant to this one though!)](https://github.com/AdoptOpenJDK/openjdk-build/issues/1581)

**Comment by lumpfish** [comment]
I managed to miss the fact that a milestone 2 build for ppcle was missing entirely (build failure)  by relying too heavily on TRSS - if a build fails you don't see test failures, but you're not alerted to the build failure either - there is simply no row for that platform.

Might it be possible to get TRSS to somehow also show build failures in a pipeline (at least include a row for the broken build)?

**Comment by sxa** [comment]
As a clarification to the previous comment - this is an OpenJ9-specific issue since they create their release branches weeks before release, and double-fix critical fixes for the release, and produce milestone builds from that. The branch in question had the [JITserver support enabled on Linux/ppc64le](https://github.com/AdoptOpenJDK/openjdk-build/issues/1670) (we use the same scripts for nightly and release builds so it got enabled for both) but had not had the [fix to remove the protobuf dependency](https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/281) removed, and the Linux/ppc64le build machines did not have the protobuf package installed as it had never been required for the nightly builds on that platform. Solution that @lumpfish is implementing is to port the ["I don't need protobuf" patch](https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/281) across and rebuild all platforms to ensure consistent SHAs across releases, which will also mean that the fix for the [large JREs on Windows](https://github.com/AdoptOpenJDK/openjdk-build/issues/1670) will be included, but will delay the releases.[

**Comment by sxa** [comment]
[Ubuntu-3/ppc64le test machine](https://ci.adoptopenjdk.net/computer/test-osuosl-ubuntu1604-ppc64le-3/) had a connection issue [during testing](https://ci.adoptopenjdk.net/job/Test_openjdk14_j9_sanity.openjdk_ppc64le_linux/59/console).

**Comment by andrew-m-leonard** [comment]
In doing the OpenJ9 Triage I felt as though I was "lacking":
1. Overall visual view of the current "status" of a given pipeline-job. eg.All build statuses? TRSS is great for test view, but reliance on that can lead to missing build failures.
2. I can't tell what bits of a pipeline job are still running... I discovered a pLinux test job had hung 7 hours earlier, a better "overall status" view would have made spotting this easier.
3. Being able to check if jobs are "Waiting for available executor", then be able to determine there are actually "online" executors but they're just running other jobs, as opposed to all available executors "offline"!?

**Comment by sxa** [comment]
>     1. Overall visual view of the current "status" of a given pipeline-job. eg.All build statuses? TRSS is great for test view, but reliance on that can lead to missing build failures.

That's what I personally use the [failing builds tab](https://ci.adoptopenjdk.net/view/Failing%20Builds/) for - it lets you see at a glance what has had a complete failure

>     2. I can't tell what bits of a pipeline job are still running... I discovered a pLinux test job had hung 7 
hours earlier, a better "overall status" view would have made spotting this easier.

While it could do with some updating the [Build Monitor](https://ci.adoptopenjdk.net/view/Build%20Monitor/) page lets you see that sort of thing as I would expect it to show in red. 

>     3. Being able to check if jobs are "Waiting for available executor", then be able to determine there are actually "online" executors but they're just running other jobs, as opposed to all available executors "offline"!?

So they are all the things in the "build queue" section of the jenkins home page. While it's unfortunate that the console log doesn't let you click the link, if something is blocked on a tag you can go to a link such ashttps://ci.adoptopenjdk.net/label/ci.role.test&&hw.arch.x86&&sw.os.linux&&sw.tool.docker/ which would show you what machines https://ci.adoptopenjdk.net/job/Test_openjdk8_j9_sanity.external_x86-64_linux_openliberty-mp-tck/262/console is waiting for

Given the above I'm conscious of the fact we have a lot of monitoring tools available for this sort of thing, but if they are unsuitable (rather than us thinking they don't exist) then we should look to make them more suitable. Bear in mind that these views have been created by people in the same situation in order to make the things your describing easier than they would be.

**Comment by lumpfish** [comment]
As far as I can see, if TRSS had a row for all platforms in a pipeline for which builds were submitted regardless of whether the build completed successfully then we would have a single toold view of what is required to triage a pipeline.  If one has to navigate through multiple tools there is a higher likelihood something will be forgotten (as I just demonstrated).

**Comment by lumpfish** [comment]
The 'OpenJ9 specific issue' was only OpenJ9 specific in as much as they create milestone builds.  The real issue is that the traiger (me) failed to notice that a platform within the pipeline failed to build by relying on the TRSS view.  That could have been the case for a milestone, nightly or release build.  So the triager needs to know to also use some other tool to provide that information or (I think ideally) TRSS enhanced to provide it.

**Comment by sxa** [comment]
I think it's two issues
1. The fact we had a real problem that was likely not visible to OpenJ9's CI (and it was invisible outside the milestone builds at AdoptOpenJDK as we don't regularly build the pre-release OpenJ9 branches)
2. The fact that it hadn't been spotted when triaging the milestone builds that showed it

For the second of those, I haven't personally found checking multiple locations too problematic (maybe because I usually have multiple screens!) but I appreciate that others may. This sounds like a valid [request against TRSS](https://github.com/AdoptOpenJDK/openjdk-test-tools) if they're willing to accept it (it is a bit more scope creep compared to providing just test results!) although I'm not sure everything in Andrew's list should go in there :-)

**Comment by andrew-m-leonard** [comment]
1. Agree "Failing builds tab" is useful, I do use that.
2. The Build Monitor view I find not very helpful! It's not Pipeline scentric, can't see the wood for the trees..... a) I "just" want to see the jobs for my pipeline build of interest, ideally in a left to right "tree" view b) I can see the status are tests Queued/Running?

I'm imagining something like this:
![image](https://user-images.githubusercontent.com/31470007/79469942-b3993280-7ff8-11ea-90ea-9fb666c3ce63.png)

**Comment by andrew-m-leonard** [comment]
BlueOcean tries to do this, but not very well!

**Comment by sxa** [comment]
I think this topic possibly needs to be split off into a separate discussion from the retrospective as it's not purely related to the release. Maybe it's a contribution you could make to BlueOcean though @andrew-m-leonard ;-)

**Comment by andrew-m-leonard** [comment]
I don't get on with BlueOcean (can you tell!)
I feel an extension to TRSS, or something along side it would work, and most likely not too difficult to put together in that framework

**Comment by andrew-m-leonard** [comment]
So didn't realize you could do this:
https://ci.adoptopenjdk.net/label/ci.role.test&&hw.arch.x86&&sw.os.linux&&sw.tool.docker/
Would be nice to generate this as a button/link in a build description when a job is scheduled maybe...?
Would also be nice if that view would display execution status of each, rather than having to click through to each.....

**Comment by andrew-m-leonard** [comment]
> I think this topic possibly needs to be split off into a separate discussion from the retrospective as it's not purely related to the release. Maybe it's a contribution you could make to BlueOcean though @andrew-m-leonard ;-)

It feels related to me as the 3 hours this morning triaging the release made me think of these....?!

**Comment by andrew-m-leonard** [comment]
Thought i'd mention something positive!
"Concurrent Pipelines" really worked well, no hanging around

**Comment by sxa** [comment]
> It feels related to me as the 3 hours this morning triaging the release made me think of these....?!

True, but also should be true for whoever is checking the nightly builds on a daily basis - I don't think the fact this is a release week has resulted in that effort being particularly bad, unless it's due to the amount of build concurrency going on, or do you disagree?

**Comment by andrew-m-leonard** [comment]
I think having a combination of BlueOcean view on one screen, and TRSS on the other screen helps me:
![image](https://user-images.githubusercontent.com/31470007/79474097-e8f44f00-7ffd-11ea-8b1f-e795d3bfce56.png) 
![image](https://user-images.githubusercontent.com/31470007/79474406-3ffa2400-7ffe-11ea-85cd-1295eda38329.png)

**Comment by smlambert** [comment]
TRSS does have a view that shows failed builds and one that shows all children of the pipeline (not just tests, though it is a Test Results Summary Service, so it is test-centric and we had reserved the Grid view for test results).

See here:
<img width="1304" alt="Screen Shot 2020-04-16 at 5 42 21 PM" src="https://user-images.githubusercontent.com/2836948/79509803-3c6c9a00-800a-11ea-8025-6d1d9e6c0cfc.png">

- clicking on the link indicated by the blue oval shows **all** children of the pipeline and their status (build and test jobs)

- clicking on the link in the Build status column called **Failed Builds** (indicated by red oval and star) will show failed builds... 

Here is a view when I click on that Failed Builds link for pipeline 983 (which is I believe the m2 one from April 15th showing the ppc64le build failures):
<img width="1311" alt="Screen Shot 2020-04-16 at 5 41 30 PM" src="https://user-images.githubusercontent.com/2836948/79510037-a38a4e80-800a-11ea-8c63-bcc1fb3edc41.png">

**Comment by smlambert** [comment]
Re: https://github.com/AdoptOpenJDK/TSC/issues/151#issuecomment-614701763, see the build description of test jobs ([LABEL used + machine that the label resolved to](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/buildenv/jenkins/JenkinsfileBase#L580-L585) for running the test on)

<img width="742" alt="Screen Shot 2020-04-16 at 5 57 28 PM" src="https://user-images.githubusercontent.com/2836948/79510717-ded94d00-800b-11ea-9fa5-9b4e2071cfc7.png">

**Comment by andrew-m-leonard** [comment]
Installer job still seems to be a bottleneck, and hung on one job:
https://ci.adoptopenjdk.net/job/build-scripts/job/release/job/create_installer_mac/2843/console

**Comment by andrew-m-leonard** [comment]
Re-run as Grinder button never seems to work for me!?

**Comment by andrew-m-leonard** [comment]
No artifact links on builds makes finding correct link to re-run Grinder hard... : https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk14u/job/jdk14u-aix-ppc64-openj9/39/
no link?

**Comment by sxa** [comment]
> Installer job still seems to be a bottleneck, and hung on one job:
> https://ci.adoptopenjdk.net/job/build-scripts/job/release/job/create_installer_mac/2843/console

"Still seems to be bottleneck" << I'm not aware we've ever seen the mac installers blocked before - have you seen that previously Andrew?

**Comment by sxa** [comment]
> 
> 
> No artifact links on builds makes finding correct link to re-run Grinder hard... : https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk14u/job/jdk14u-aix-ppc64-openj9/39/
> no link?

This will be as a result of reducing further the number of builds we keep in order to reduce disk space a few months back. We have a little more now so could relax that a bit now.

**Comment by andrew-m-leonard** [comment]
ahhh, thanks @simon for the tip.
it’s because the artifacts are not stored in the build job, but in the upstream Pipeline job:
https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk11-pipeline/865/artifact/target/windows/x64/openj9/
I suspect that’s why the re-run Grinder does not work as well

**Comment by andrew-m-leonard** [comment]
> > Installer job still seems to be a bottleneck, and hung on one job:
> > https://ci.adoptopenjdk.net/job/build-scripts/job/release/job/create_installer_mac/2843/console
> 
> "Still seems to be bottleneck" << I'm not aware we've ever seen the mac installers blocked before - have you seen that previously Andrew?

never, looking at the exception when I aobrted it, looks like it had an issue copying upstream artifact

**Comment by sxa** [comment]
> Re-run as Grinder button never seems to work for me!?

Can you provide an example of a link that doesn't work and what the symptoms are?

**Comment by andrew-m-leonard** [comment]
> > Re-run as Grinder button never seems to work for me!?
> 
> Can you provide an example of a link that doesn't work and what the symptoms are?

https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/2843/parameters/
```
ERROR: Failed to copy artifacts from build-scripts/jobs/jdk11u/jdk11u-windows-x64-openj9 with filter: **/*.tar.gz,**/*.tgz,**/*.zip,**/*.jar
Finished: FAILURE
```
As pointed out above artifacts are not in the given job, but are in the upstream Pipeline job one level further up...

**Comment by sxa** [comment]
(IMHO I still feel it would be preferable to move general triage gripes that aren't specific to this release into a separate issue to keep it easy to track the real issues that affected this release in particular)

**Comment by sxa** [comment]
Another point: The mac installer job got stuck and had to be aborted, this meant that the pipeline job (used for publishing artefacts) only had a `.zip` file and not the macos `.pkg`. We have re-run the job, but that doesn't automatically go back into the pipeline artefacts. There are multiple options for this (some of which are rendered impossible already but I'll cover them for completeness:

1. Manually upload the artefact (impossible as `create_installer_mac` has already been re-run and it only stores one artefact, so the release one has already been deleted from the artefact history
2. Re-run the `create_installer_mac` again with the build, and store the artefacts (lock the job, copy them elsewher,e maybe even copy them into the pipeline artefacts directory on the jenkins server
3. Re-run the complete macos pipeline again, although we'd potentially have to publish the existing artefacts from the first pipeline, and then the new macos pipeline publish would "fill in the blanks" (This might be undesirable as the original zip and the new pkg could end up from different builds unless we removed the zip artefacts from the first pipeline jobs on the jenkins server.

Attempting 2 was problematic as the installer job can't seem to use the artefacts from the pipeline (possibly because it doesn't squash directory paths so the artefacts aren't where it expects them) therefore we are going to use option 3

**Comment by smlambert** [comment]
**Proposed action**: changes to pipeline scripts (including moving where artifacts are stored) need to be widely communicated and part of the review of those changes need to scrutinize other pipeline consumers to see what will be affected (those affected currently installers / test jobs, rerun in Grinder link).

**Proposed action**: finish [test doc update](https://github.com/AdoptOpenJDK/openjdk-tests/pull/1556), UX session to watch and understand how triagers currently work, and create tutorials for 'best practices' with current tools.  Specifically Rerun in Grinder link is there to _help prepopulate the fields_ in the Grinder.  The user should still have an understanding of what the parameters mean, so they can tune and tweak to suit their purpose.  

I also agree that we should use this issue to collect a list of actions that may help the next release rather than a gripe session (which is rather disheartening, I personally feel our tools and systems have improved tremendously from release to release, even though there is still a lot of improvements to come).

**Comment by smlambert** [comment]
**Proposed action**: as part of the pre-release checklist, release manager (or delegate) ensures that both hotspot and openj9 test jobs for both implementations and all versions that are part of the release have been fully triaged (attempt to 'run green' always, but very least the week prior to the release)

**Comment by gdams** [comment]
aarch64 shenandoah tag was empty, It seems that we have to delete the tag from github and then re-run the mirror job specifying `move-to-next`.

**Comment by sxa** [comment]
> aarch64 shenandoah tag was empty, It seems that we have to delete the tag from github and then re-run the mirror job specifying `move-to-next`.

Was that the same issue we had on the main jdk8 repo at the start?

**Comment by gdams** [comment]
> Was that the same issue we had on the main jdk8 repo at the start?

@sxa it was indeed

**Comment by sxa** [comment]
> **Proposed action**: changes to pipeline scripts (including moving where artifacts are stored) need to be widely communicated and part of the review of those changes need to scrutinize other pipeline consumers to see what will be affected (those affected currently installers / test jobs, rerun in Grinder link).

@smlambert Were there any particular changes that we've already made that you're thinking of here - we haven't moved anything related to artifact storage yet and the number of artifacts we keep was [reduced  to 1](https://github.com/AdoptOpenJDK/openjdk-build/pull/1161) on the build jobs some time ago (which can now be relaxed a bit, especially since we're doing concurrent pipelines (saying that, we apparently [switched it back up to 10](https://github.com/AdoptOpenJDK/openjdk-build/pull/1282) temporarily during one previous release, so that's an option ..), but we should still move to external storage) or was there something else that was a concern in this area?

**Comment by andrew-m-leonard** [comment]
website/apiv3 repeatedly does not pickup newly published openj9 releases...
11.0.7+10_openj9-0.20.0 was published nearly 3 hours ago, and website still not showing it....?
Raised issue: https://github.com/AdoptOpenJDK/openjdk-api-v3/issues/184

**Comment by smlambert** [comment]
That proposed action was triggered based on an attempt to triage this release when we tried to rerun in a grinder we could not do it as artifacts were no longer present.  I think reducing the number of artifacts down to 1 is suboptimal, perhaps a few days padding is best... 

It is fine to delete the artifacts if: 1) the artifacts are available from somewhere else (published and available from API) AND 2) consumer is written to pull from somewhere else (test jobs can, but are installers able?).  

I believe we did have a case during this release where @gdams wanted to run grinders against a binary that had not yet been published, so was not available through API, but also not attached as an artifact to the upstream job.

Also, seeing triagers click on Rerun in Grinder link and not have it be possible, because the original job no longer has artifacts... we need triagers to either know that the upstreams are purged and they then manually adjust the parameters (to customized and point to the API URL, or to nightly if its a recurring issue), or we keep a few more days of artifacts for SDK_RESOURCE=upstream and find a happy compromise.

**Comment by aahlenst** [comment]
We have standalone_create_installer_x (where x stands for each platform) jobs that should produce installers/packages via GitHub binaries. At least the Linux variant is fully operational.

I‘m wondering whether it might make sense to decouple the installer/package production. Trigger an event in Jenkins after packages have been promoted that kicks off those jobs. Reduces the risk of failures and we do not need full rebuilds if something goes wrong.

**Comment by smlambert** [comment]
https://adoptopenjdk.slack.com/archives/CLCFNV2JG/p1587546262171100 discussion about https://github.com/AdoptOpenJDK/openjdk-build/pull/1695  

Suggestion is to have a test that loops through executables and ensures that they are signed

Other notes:
Related: https://ci.adoptopenjdk.net/job/build-scripts/job/release/job/sign_build/. <- essentially if this is failing its BAD.  We can assume that the release manager does not know to monitor those jobs?

Related: https://github.com/AdoptOpenJDK/openjdk-tests/issues/1421

**Comment by aahlenst** [comment]
With https://github.com/AdoptOpenJDK/openjdk-build/issues/1699 we have another incident of archives including material that should not be included.

It might be worthwhile to create a baseline of archive contents per major release and compare after every build. Even a size comparison with some wiggle room might be enough.

**Comment by sxa** [comment]
> I‘m wondering whether it might make sense to decouple the installer/package production.

@aahlenst Just to clarify, which problem is this seeking to address? If we are able to run the installer/package separately if required when the packaging step fails but I'd be reluctant to remove it from the main pipelines (if for no other reason than it adds complexity for those running builds)

**Comment by DanHeidinga** [comment]
Proposed in `OpenJDK macos and AIX quarterly builds have too much content` #9318:
> With the IBM SDK, we had an automated job that checked that the SDK hadn't grown in size by more than small delta and flagged new files / removed files.
>
> Is it worth adding a similar job to the Adopt pipelines to catch the SDK binaries haven't become too large?

**Comment by smlambert** [comment]
Stepping back and looking at the types of issues that we tend to see during any given release, I think it would be useful for us to:

1. Define what is a release (the [Release guide](https://github.com/AdoptOpenJDK/openjdk-build/blob/master/RELEASING.md#adoptopenjdk-release-guide) is a good start, let's ensure every activity of component is captured in it, including creating installers, publishing docker images, writing release notes, etc).

1. Ensure that there is an overarching workflow/pipeline for each of the items in a release definition.

1. Verify each step in the flow.  Currently we have this in place for the 'build component'.  We build, then we run testing, but need it for every other component (docker, installers, documentation, website, etc).

I think the 2nd point will be useful from the perspective of knowing when a release is 'done'.  It is not just when the SDKs have been built/tested/published, but must also include the other components.  I also think that visualizing this flow, will make it easier to onboard others and get more help.  I will take a first try at this definition/flow this week.

**Comment by 0xdaryl** [comment]
I think updating the AArch64 build compiler from 7.4 -> 7.5 a week before building the final release was a bit risky (eclipse/openj9#9159).  Although likely a straightforward upgrade since OpenJ9 was already being built on 7.5 on other Linux platforms and the fact that AArch64 was only an "early access" release, it was a little close to the edge.

Perhaps the very fact that this was an early access release meant that the upgrade could have been deferred.

Things seem to have worked out, but there were a couple of failures in the final week's push to stabilize on AArch64 where that compiler upgrade was investigated as a possible cause.

**Comment by sxa** [comment]
@0xdaryl I wasn't aware of that, but AdoptOpenJDK was using gcc 7.5 on our aarch64 build machines and had been [since February](https://github.com/AdoptOpenJDK/openjdk-build/issues/1508) I believe so that had no impact from our perspective. Arguably being backlevel on the compiler was a fault on the OpenJ9 side and updating it reduces the risk to the releases by having the OpenJ9 machines brought in sync with the release build systems.

**Comment by adamfarley** [comment]
Note: Any unresolved actions have been folded into the next retrospective for review. [Link.](https://github.com/adoptium/adoptium/issues/77)

If any have been unintentionally missed, feel free to add them.
